### PR TITLE
[Serve] Type hints for API

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
Type hints provide a better experience for the users of our API.  They can use their own static type checker to prevent errors.

Config dicts are given the type `Dict[str, Any]`.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing.  (ran `test_api.py`) Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/serve/api.py [hunk 1]
```diff
@@ -235,11 +238,11 @@ def get_backend_config(backend_tag):
 
 
 @_ensure_connected
-def create_backend(backend_tag,
-                   func_or_class,
+def create_backend(backend_tag: str,
+                   func_or_class: Callable,
```
Comment (simon-mo) [inline python/ray/serve/api.py:242]: `Union[Callable, Type[Callable]]`. `Callable` for functions and `Type[Callable]` for classes that implement `__call__`

File: python/ray/serve/api.py [hunk 2]
```diff
@@ -235,11 +238,11 @@ def get_backend_config(backend_tag):
...
                    *actor_init_args,
```
Comment (simon-mo) [inline python/ray/serve/api.py:243]: `*actor_init_args: Any`

File: python/ray/serve/api.py [hunk 3]
```diff
@@ -235,11 +238,11 @@ def get_backend_config(backend_tag):
...
                    ray_actor_options=None,
```
Comment (simon-mo) [inline python/ray/serve/api.py:244]: `Optional[Dict]`

File: python/ray/serve/api.py [hunk 4]
```diff
@@ -235,11 +238,11 @@ def get_backend_config(backend_tag):
...
-                   config=None):
+                   config: Dict[str, Any] = None) -> None:
```
Comment (simon-mo) [inline python/ray/serve/api.py:245]: `Optional[Dict[str,Any]]`

**Review by simon-mo — APPROVED** [review]
LGTM with few minor changes. Please remove the added binary files.

File: python/ray/serve/api.py (hunk 4 above)
Comment (architkulkarni) [reply python/ray/serve/api.py]: All fixed, thanks for the careful check!

**Pull request merged**